### ISSUE TITLE
TinyDNS backend not answering location restricted zones

### ISSUE THREAD

**Issue by rsevero** [PowerDNS/pdns#13315]
- [X] This is not a support question, I have read [about opensource](https://www.powerdns.com/opensource.html) and will send support questions to the IRC channel, [Github Discussions](https://github.com/PowerDNS/pdns/discussions/) or the mailing list.
- [X] I have read and understood the ['out in the open' support policy](https://blog.powerdns.com/2016/01/18/open-source-support-out-in-the-open/)

 - Program: Authoritative 
 - Issue type: Bug report

### Short description

PowerDNS with TinyDNS backend doesn't answers as authoritative for zones which the regular "name server" lines (the ones that begin with a dot) have location restrictions.

The zone data has location restrictions and PDNS respectes it accordingly but if I include the same location restriction on the DOT line, PDNS stops answering for this zone.

The original TinyDNs answers for zone for which their DOT lines have location restrictions.

### Environment

 - Operating system: 33~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC with kernel 6.2.0-33-generic
 - Software version: PowerDNS Authoritative Server 4.8.2
 - Software source: PowerDNS repository

### Steps to reproduce

1. Start a PDNS server with a TinyDNS backend on a machine with one IP address: 192.168.1.1.
2. Create a _data.cdb_ file from the below _data_ file:
```
%in:192.168.1.1

.168.192.in-addr.arpa:192.168.1.1:a:::in

^1.1.168.192.in-addr.arpa:test.example.com:::in
```
3. Restart pdns server
4. Query the pdns server for the PTR record for 192.168.1.1: `dig -x 192.168.1.1 @192.168.1.1 +norecurse`

### Expected behaviour

Get the PTR info for 192.168.1.1 as in 
```
; <<>> DiG 9.18.12-0ubuntu0.22.04.3-Ubuntu <<>> -x 192.168.1.1 @192.168.1.1 +norecurse
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 17433
;; flags: qr aa; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;1.1.168.192.in-addr.arpa.      IN      PTR

;; ANSWER SECTION:
1.1.168.192.in-addr.arpa. 86400 IN      PTR     test.example.com.

;; Query time: 0 msec
;; SERVER: 192.168.1.1#53(192.168.1.1) (UDP)
;; WHEN: Fri Sep 29 13:04:24 -03 2023
;; MSG SIZE  rcvd: 83
```

### Actual behaviour

Got no PTR info:
```
; <<>> DiG 9.18.12-0ubuntu0.22.04.3-Ubuntu <<>> -x 192.168.1.1 @192.168.1.1 +norecurse
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 34121
;; flags: qr; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;1.1.168.192.in-addr.arpa.      IN      PTR

;; Query time: 0 msec
;; SERVER: 192.168.1.1#53(192.168.1.1) (UDP)
;; WHEN: Fri Sep 29 12:58:42 -03 2023
;; MSG SIZE  rcvd: 53
```

### Other information

If I simply remove the location restriction **from the DOT line only** PDNS starts to provided the expected PTR info. Working _data_ file:
```
%in:192.168.1.1

.168.192.in-addr.arpa:192.168.1.1:a

^1.1.168.192.in-addr.arpa:test.example.com:::in
```

Note: it is necessary to change the data/data.cdb files **AND** restart pdns to make it answer for the mentioned zone.

**Comment by omoerbeek** [comment]
Did you read https://docs.powerdns.com/authoritative/backends/tinydns.html#location-and-timestamp-support ?

**Comment by rsevero** [comment]
> Did you read https://docs.powerdns.com/authoritative/backends/tinydns.html#location-and-timestamp-support ?

Yes. Several times. It says nothing about location restrictions being supported in some types of DNS records and not being supported in others as far as I can tell. Or does it?

**Comment by rsevero** [comment]
> zone-cache-refresh-interval=0

Just did. The behaviour didn´t change, the problem remains. Tried with the location restriction on and off the DOT data line.

**Comment by Habbie** [comment]
Ok, I had a look. Backends (like the tinydnsbackend) do not get information about the remote (the client) for the internal SOA lookup. While we aim to be as compatible with tinydns as we can, this one will be quite involved to fix, so I think we might just document this limitation.

**Comment by rsevero** [comment]
> Ok, I had a look. Backends (like the tinydnsbackend) do not get information about the remote (the client) for the internal SOA lookup. While we aim to be as compatible with tinydns as we can, this one will be quite involved to fix, so I think we might just document this limitation.

I imagined that the reason was something like that. Thanks for your help and attention.

Here is a suggestion of text to be included in th  end of the [Useful implementation Notes](https://doc.powerdns.com/authoritative/backends/tinydns.html#useful-implementation-notes) section, just before the last sentence:

> PowerDNS and TinyDNs also handle name server records (the dot prefixed lines in _data_ files) differently. In TinyDNs it is possible to apply location restriction to dot lines and the restriction is respected accordingly. In PowerDNS any location restricted dot line (and it's accompaning SOA record) is simply ignored making PowerDNS to not assume authority over the corresponding zone. This happens because in PowerDNS, backends in general, and tinydns in this case, do not get information about the remote (the client) for the internal SOA lookup.
> 
> If you want to apply location restrictions to zones, apply the location restrictions on the rest of the zone data but not on the name server line (dot prefixed line).

**Comment by Habbie** [comment]
Oh thanks! This will be useful :)

**Comment by mind04** [comment]
Related #5512 #7057
